### PR TITLE
Add gasless EVM proposal

### PR DESCRIPTION
**Pull request by garious**
#### Problem

Applications targeting other blockchains use audited smart contracts implemented in Solidity. Running that same application targeting Solana requires migrating the smart contract to Rust and the Sealevel parallel programming model. Creating the same smart contract in Solana requires mechanical transformations of high-level Solidity constructs to lower-level Rust equivalents. The process is tedious and allows for human error, requiring a second audit of the new code. Automating the process would save time and alleviate the need for an audit.

#### Summary of Changes

Add a design proposal

### CODE REVIEW

**Comment by aeyakovenko** [comment]
That's awesome.  I think we can deal with the instruction limit by AOT bpf into x86 at the loader.  So I am not to worried about it.  In theory it should be only 4x bigger :)

**Comment by aeyakovenko** [comment]
Reallocate via a CPI seems doable, since the program is aware of the length change.

File: docs/src/proposals/gasless-evm.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+solana-evm lib-hello.json HelloWorld "Hello world!" -a hello-world.json
+solana-evm lib-hello.json update "Goodbye cruel world!" -a hello-world.json
+```
+
+### EvmInstruction
+
+```
+///   0. `[]` Clock sysvar
+///   1. `[]` Contract library containing method
+///   *. `[writable]` Contract accounts
```
Comment (garious) [inline docs/src/proposals/gasless-evm.md:45]: These can be a mix of writable/readonly/signer. How to express that?

File: docs/src/proposals/gasless-evm.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+accounts they may access, such that Sealevel may execute independent contracts
+in parallel. Instead of requiring "gas", map EVM bytecode to BPF so that
+contracts are subject to the BPF instruction limit. Support Solidity method
+calls from arbitrary Solana programs and executing Solana instructions from
+Solidity methods.
+
+### Developer Experience
+
+```bash
+solc --bin sourceFile.sol > lib-hello.bin
```
Comment (garious) [inline docs/src/proposals/gasless-evm.md:29]: This isn't the command we want, but all the right information is in that file. Instead, we want a binary with the binaries of each contract and some header to look up a contract's bytecode.

File: docs/src/proposals/gasless-evm.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+`EvmInstruction::Invoke` that contains a hash of the contract name, and a
+serialized string in `EvmInstruction::args`.
+
+When the EVM starts executing the `HelloWorld` constructor, it uses the first
+account from the account store to host all member variables. Since the user
+controls what accounts are passed in, the account must also hold type
+information, including the address of the contract library, and a hash of the
+contract name.
+
+When the `update()` function is called, the EVM reallocates memory for the
```
Comment (garious) [inline docs/src/proposals/gasless-evm.md:83]: Is this realloc possible today?

File: docs/src/proposals/gasless-evm.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+```bash
+solana-keygen new -o hello-world.json
+solana-evm lib-hello.json HelloWorld "Hello world!" -a hello-world.json
+solana-evm lib-hello.json update "Goodbye cruel world!" -a hello-world.json
+```
+
+### EvmInstruction
+
+```
+///   0. `[]` Clock sysvar
```
Comment (garious) [inline docs/src/proposals/gasless-evm.md:43]: Might be able to push this into the contract accounts. If you don't use TIMESTAMP, then you shouldn't have to burn 32 bytes of transaction space passing this in.

**Comment by hendrikhofstadt** [comment]
Awesome! How are you planning to deal with precompiles/builtins/syscalls? The EVM has quite some so called precompiles (for signature verification, hashing, etc) which are calling optimized libraries outside the EVM.

**Comment by garious** [comment]
@hendrikhofstadt, initially, that'll be done with cross-program invocation (CPI), but those calls won't be parallelized on the GPU. To do that, we'll need to change CPI to asynchronously call into the runtime so that we could batch them, offload to the GPU, and then resume the caller.

**Comment by aeyakovenko** [comment]
@garious we could just have the user specify ahead of time what the expected CPI's will look like and batch/execute them before they land into the runtime.

**Comment by garious** [comment]
@aeyakovenko, I was thinking the opposite, that we'd no longer need to special case sigverify stage. Instead, every contract can start with an implicit sigverify instruction, and that the same runtime mechanism would batch and offload that work.

**Comment by aeyakovenko** [comment]
@garious for the general case, 100% agreed.

But a ton of crypto operations can be transaction local instructions which do not require any state context to execute.  Those calls validators can execute out of order, even optimistically when forwarding.

File: docs/src/proposals/gasless-evm.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+///   1. `[]` Contract library containing method
+///   *. `[writable]` Contract accounts
+Invoke { method_id: Hash, args: Vec<u8> },
+```
+
+### Contract to account mapping
+
+A Solana instruction includes a list of public keys corresponding to any account
```
Comment (0b01) [inline docs/src/proposals/gasless-evm.md:51]: Implement storage as a hash table with constant size buckets, whose entries are 64-byte (u)int256 key:value pairs. When it gets 75% full, create a new account [program derived address] (https://docs.solana.com/implemented-proposals/program-derived-addresses) and double the storage. Then when SLOAD is called, if `hash % n` isn't found in the bucket, try `hash % (n/2)` and move the entry to `hash % n`.

File: docs/src/proposals/gasless-evm.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+### Runtime data as sysvars mapping
+
+The EVM has special instructions to access runtime data, such as TIMESTAMP. To
+access the timestamp, the EVM will query the account store for
+`sysvar::clock::id()`.
+
+### Going gasless
+
+Ethereum uses *gas* to ensure a smart contract terminates. In Solana, a
+contract is allowed to run up to `N` BPF instructions, where `N` is set by the
```
Comment (0b01) [inline docs/src/proposals/gasless-evm.md:95]: Ethereum allows 1024-deep cross-contract calls. Will this hit `N` BPF instruction limit?

File: docs/src/proposals/gasless-evm.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+protocol. Therefore, when the Solidity contract sends gas to other contracts,
+the Solana EVM simply ignores it.
+
+### Calling Solidity methods from Solana programs
+
+Since Solidity methods are invoked with Solana instructions, those instructions
```
Comment (0b01) [inline docs/src/proposals/gasless-evm.md:101]: It is possible to map solana instructions to eth abi formatters.

File: docs/src/proposals/gasless-evm.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+may be invoked from any Solana program using its cross-program invocation
+mechanism.
+
+### Executing Solana instructions from Solidity methods
+
+The EVM is extended with a built-in function `invoke` that calls the
```
Comment (0b01) [inline docs/src/proposals/gasless-evm.md:107]: Also need to handle REVERT which restores storage state to before any opcode executes.

File: docs/src/proposals/gasless-evm.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+cross-program invocation mechanism.
+
+### Long-running smart contracts
+
+Long-running smart contracts are not supported by the gasless EVM. Instead, the
+smart contract would need to be executed on a traditional EVM, which is outside
```
Comment (0b01) [inline docs/src/proposals/gasless-evm.md:113]: Add timeout

File: docs/src/proposals/gasless-evm.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+the scope of this proposal.
+
+### Technical risks
+
+* 256-bit EVM instructions may hit the BPF instruction limit so quickly that
+  only the smallest of Solidity smart contracts would work
+* Solana's UDP packets must stay under the MTU limit, 1,500 bytes. Complex
+  Solidity contracts may implicitly reference many accounts and therefore
+  exceed that limit.
```
Comment (0b01) [inline docs/src/proposals/gasless-evm.md:122]: That's about ~46 `uint256`s. What are the UDP packets used for? Cross contract calls can be implemented natively without UDP.

File: docs/src/proposals/gasless-evm.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+### Long-running smart contracts
+
+Long-running smart contracts are not supported by the gasless EVM. Instead, the
+smart contract would need to be executed on a traditional EVM, which is outside
+the scope of this proposal.
+
+### Technical risks
+
+* 256-bit EVM instructions may hit the BPF instruction limit so quickly that
```
Comment (0b01) [inline docs/src/proposals/gasless-evm.md:118]: What's the limit?

File: docs/src/proposals/gasless-evm.md (hunk 2 above)
Comment (0b01) [reply docs/src/proposals/gasless-evm.md:29]: constructor, runtime, and abi info

File: docs/src/proposals/gasless-evm.md (hunk 6 above)
Comment (garious) [reply docs/src/proposals/gasless-evm.md:95]: N is currently 200,000. Wether 1024 calls hits that limit will depend on the function call implementation.

File: docs/src/proposals/gasless-evm.md (hunk 10 above)
Comment (garious) [reply docs/src/proposals/gasless-evm.md:122]: We use UDP to send transactions and the transaction must include the public keys of any accounts that the smart contract requires.

File: docs/src/proposals/gasless-evm.md (hunk 3 above)
Comment (jackcmay) [reply docs/src/proposals/gasless-evm.md:83]: Generalized realloc is not supported, just `CreateAccount`

**Pull request closed without merging**